### DISCUSSION TITLE
Auto-completion. IDE integration and up to date documentation

### DISCUSSION THREAD

**Discussion by ScreamZ** [Moddable-OpenSource/moddable#1262] [Getting started]
Hi,

I just found moddable SDK which is nice, I'm guessing is there any up-to-date documentation? I can see there are a lot of examples, especially with drivers that are not updated in the documentation folder.

Also, I would like to know if there exists some good IDE integration, especially for auto-completion with imports.
I can see that typescript support is experimental, but there are no @types/ typings and things like that.

Is there any way I could help and eventually contribute to improving all those subjects, I found it very sad that this technology seems so promising but it's too opaque for beginners to jump in.

Thank you,
Andréas

**Comment by tve** [comment]
Hi @ScreamZ! Your questions are probably best moved to the Q&A section of the github discussions...
You can find the general documentation in the `documentation` directory (as you already did) and many of the drivers and libraries are "documentation by example".

I have not found a good integration with vscode per-se, but it works reasonably well after applying some elbow grease. There are TS typings in the `typings` directory. They are not complete but go quite far. I've been adding missing ones as I go along. The main difficulty with TS is that the `tsconfig.json` for the Moddable XS compiler gets auto-generated from the manifest of a project but that's not useful for vscode (or other IDE). So you have to manually construct a parallel `tsconfig.json` for the IDE. I can post what I have but it will need tweaking and others have also posted their tsconfig in the discussions section.

I totally agree with you that the initial learning curve is steep, especially if you've used something like MicroPython or Espruino, but the core is good so it's worth it in the end.

**Comment by phoddie** [comment]
>  is there any up-to-date documentation

The [documentation](https://github.com/moddable-OpenSource/moddable/tree/public/documentation) directory is kept up-to-date. No documentation is perfect. Feedback and improvements are welcome.

Also, there's an easier to nativate and read version of the same [documentation on the Moddable web site](https://www.moddable.com/documentation/readme).

Our [TypeScript declarations](https://github.com/Moddable-OpenSource/moddable/tree/public/typings) are maintained by the community. They are quite usable. As @tve notes, you do need to add them to your project's tsconfig.json.
